New soiltex for ctsm5.2.mksurfdata

slevisconsulting opened the pull request:
### Description of changes
Started from @mvertens branch feature/new_mksurfdata_soiltex
Pushed a few updates directly to that branch:
- Merge remote-tracking branch 'escomp/ctsm5.2.mksurfdata'
- Merge branch 'cmake_bld_branch'

From there I started my branch new_soiltex_slevis and opened this PR.

This PR continues @mvertens and @wwieder 's work to replace the old soil texture and organic raw datasets with new.

### Specific notes

Contributors other than yourself, if any:
@mvertens @wwieder @dlawrenncar @ekluzek 

CTSM Issues Fixed (include github issue #):
Resolves #1303

Are answers expected to change (and if so in what way)?
Yes because new rawdata.

Any User Interface Changes (namelist or namelist defaults changes)?
New soil raw dataset consists of two files: a mapunits file and a lookup file.

Testing performed, if any:
@olyson posted diagnostics from running with the new and the old [here](https://github.com/ESCOMP/CTSM/issues/1303#issuecomment-1099208099) and no red flags were raised.

slevisconsulting commented:
@mvertens I'm adding the --hires_soitex option to the .py and .xml files, but I don't know this: 
Do we have a mesh file for this 30-second grid?

slevisconsulting commented:
> Do we have a mesh file for this 30-second grid?

@mvertens from this commit
https://github.com/ESCOMP/CTSM/pull/1732/commits/a849f9548cdf227b1db0f42263b94ffb8231004f
I gather that you used the same file to read mapunits and the mesh when working with the 30'' resolution.

@ekluzek and everyone:
I have now renamed and copied the three new files and would like comments/approval before I submit them to the repository with the ./rimport command:
| old name | new name |
|----|----|
| soiltex_mapunits_4320x2160_c220329.nc | mksrf_soitex_mapunits_5x5min_from_wise_30sec.c220329.nc |
| wise_30sec_v1_lookup2.nc | mksrf_soitex.10level.wise_30sec_lookup.c220329.nc |
| wise_30sec_v5_grid.nc | mksrf_soitex_mapunits_wise_30sec.c220330.nc |

I decided to keep the prefix "mksrf_" because it's the only identifier that indicates a CTSM raw dataset. Without the prefix, the file could be anything, while with the prefix the file is clearly intended for use with the mksurfdata tool.

ekluzek commented:
I think these are good names, I only have a couple tweaks to suggest. You can take or leave as you see fit. I'm happy with above, I just wonder about a few things.

Suggest capitalizing WISE since it's an org abbreviation
Maybe change "soitex" to just "soil" -- since we now include more soil characteristics (like PH) and not just strictly soil texture?
5x5min_from_wise_30sec is maybe confusing for the filename? Maybe just say 5x5min_WISE and inside the file you can see the metadata that says it's from the 30 second resolution file.

slevisconsulting commented:
> Suggest capitalizing WISE since it's an org abbreviation Maybe change "soitex" to just "soil" -- since we now include more soil characteristics (like PH) and not just strictly soil texture? 5x5min_from_wise_30sec is maybe confusing for the filename? Maybe just say 5x5min_WISE and inside the file you can see the metadata that says it's from the 30 second resolution file.

@ekluzek thank you. Here are the file names that I ended up with based on your suggestions:
`mksrf_soil_mapunits_30sec_WISE.c220330.nc`
`mksrf_soil_mapunits_5x5min_WISE.c220329.nc`
`mksrf_soil_lookup.10level.WISE.c220329.nc`  Here I also moved "lookup" after "soil" as done with "mapunits" and removed "30sec" because this file has resolution in the vertical but not horizontal.

slevisconsulting commented:
./rimport complete on these three files.

slevisconsulting commented on tools/mksurfdata_esmf/README:
```diff
@@ -76,6 +78,7 @@ slevis HAS RUN THESE CASES and HAS LISTED ISSUES ENCOUNTERED
 ------------------------------------------------------------
 REMEMBER TO compare against existing fsurdat files in
 /glade/p/cesmdata/cseg/inputdata/lnd/clm2/surfdata_map/release-clm5.0.18
+0) New 30-sec raw data for soil texture fails. Try requesting more mem.
```
In this README file, this is the only substantial change (others just improve the formatting):
It documents that I haven't run successfully with the new 30-sec soil-texture data.

slevisconsulting commented on tools/mksurfdata_esmf/gen_mksurfdata_namelist.py:
```diff
@@ -176,7 +176,19 @@ def get_parser():
             """,
         action="store_true",
-        dest="hres_flag",
+        dest="hres_pft",
```
Renamed hres_flag to hres_pft to distinguish it from the new hres_soitex.

slevisconsulting commented on tools/mksurfdata_esmf/gen_mksurfdata_namelist.py:
```diff
@@ -390,6 +410,10 @@ def main ():
 
+            if item.tag == 'lookup_filename':
+                new_key = f"{child1.tag}_lookup"
+                rawdata_files[new_key] = os.path.join(input_path, item.text)
```
This needed for the newly introduced soil texture lookup file.

slevisconsulting commented on tools/mksurfdata_esmf/src/mksoiltexMod.F90:
```diff
@@ -98,229 +103,270 @@ subroutine mksoiltex(file_mesh_i, file_data_i, mesh_o, pioid_o, pctlnd_pft_o, rc
+          ! Adjust sand and clay be loam if pctlnd_pft is < 1.e-6 or mapunit is 0
+          sand_o(no,:) = 43._r4
+          clay_o(no,:) = 18._r4
+          orgc_o(no,:) = 18._r4  ! TODO slevis: value for loam here?
```
@wwieder these are the model's generic values for sand and clay. What do we want for organic in that case?

Also, I haven't included a units transformation here from gC/kg (lookup file) to kg/m3 (fsurdat), which I think is what you were discussing in #1303. Did you reach a verdict?

wwieder replied:
Sam what are the organic values currently used in oceans?  I'm included to do the same here?

I'm not sure we ever decided how to handle units, let's set up a meeting with @dlawrenncar to discuss after the tutorial.  I'll try setting up a meeting.

slevisconsulting replied:
> Sam what are the organic values currently used in oceans? I'm included to do the same here?

Sounds good. So I changed to 0 here and also in cases when the lookup table returns a negative value. But not when the lookup table returns -4 (sand dunes) where I set organic = 1. This is on the low side compared to what I see from our old dataset. We can address this, too, in our meeting.

slevisconsulting commented:
This morning's meeting with @wwieder and @dlawrenncar :
1) Keep ORGC units & name unchanged when making fsurdat in mksurfdata_esmf. This allows for easier comparison between fsurdat and raw data and requires unit manipulations to happen in the land model.
2) Add CFRAG, BULK, and PHAQ to fsurdat, also with units & names unchanged.
3) Backwards compatibility:  Modify land model to use the old code when fsurdat contains ORGANIC and new code when fsurdat contains ORGC, CFRAC, BULK.

@wwieder I think the conversion formula you showed was (pls edit if incorrect):
ORGANIC = ORGC * BULK * (1 - CFRAG) * 0.01 / 0.58

I will work on (1) and (2) today.

wwieder commented:
I need to check units, but I think CFRAG is a fraction (0-100) so the equation would be 
`ORGANIC = BULK * ORGC * (100-CFRAG) * 0.01 / 0.58`

slevisconsulting commented:
> I need to check units, but I think CFRAG is a percent (0-100) so the equation would be `ORGANIC = BULK * ORGC * (100-CFRAG) * 0.01 / 0.58`

I checked it and you're right: CFRAG is in %

wwieder commented:
No I just want to check units for the whole calculation to make sure we're generating output thats similar to ORGANIC currently being used gOM/m3.

slevisconsulting commented:
> No I just want to check units for the whole calculation to make sure we're generating output thats similar to ORGANIC currently being used gOM/m3.

Here's what I'll do:
I can temporarily put both ORGC and ORGANIC (using the above conversion) on the new fsurdat file that I will generate, so that we may compare to an existing old fsurdat file.

wwieder commented:
The units should be fine, ORGC = gC/kg soil & BULK = g soil / cm^3. After converting units, the product of these = kgC/m3, as intended.  The other calculations  account for rocks and convert to gC to gOM.  Sam, I think going ahead to calculate ORGANIC on the new surface dataset like this would be helpful so we can make progress more quickly.  Good suggestion.

slevisconsulting commented on tools/mksurfdata_esmf/src/mksoiltexMod.F90:
```diff
@@ -98,237 +115,428 @@ subroutine mksoiltex(file_mesh_i, file_data_i, mesh_o, pioid_o, pctlnd_pft_o, rc
+          end do
+
+          ! TODO Rm organic before merging PR #1732 UNLESS we opt for OPTION 2
+          ! organic_o OPTION 1: as we plan to calculate organic in the CTSM
```
As a test I calculated organic_o in two different ways...
- Option 1 as we discussed this morning: calculate organic in the CTSM from orgc_o, cfrag_o, and bulk_o (i.e. after the terms have been regridded).
- Option 2 (a few lines up where orgc_o is also calculated): In this case I regrid organic_i (calculated from orgc_i, cfrag_i, and bulk_i) to organic_o. This may be the more accurate approach. It goes against this morning's idea to keep ORGC the same in fsurdat as in the raw data, but it may be more accurate because it first calculates organic_i and then regrids to organic_o rather than regridding all the terms first and then calculating organic_o.

The fsurdat files from this test are in `/glade/work/slevis/git/mksurfdata_toolchain/tools/mksurfdata_esmf`:
```
surfdata_1.9x2.5_hist_78pfts_CMIP6_1850_c220527_OPT1.nc
surfdata_1.9x2.5_hist_78pfts_CMIP6_1850_c220527_OPT2.nc
surfdata_1.9x2.5_hist_78pfts_CMIP6_1850_c220527_OPT2-1.nc (the difference file)
```
Also, the old fsurdat that I'm comparing against:
`/glade/p/cesmdata/inputdata/lnd/clm2/surfdata_map/release-clm5.0.18/surfdata_1.9x2.5_hist_78pfts_CMIP6_simyr1850_c190304.nc`
I'm happy to follow up next week. Have a nice holiday weekend!

slevisconsulting replied:
Posting the difference in ORGANIC calculated by method 2 vs. 1.
Very little difference, I think because we regrid by most dominant rather than by averaging.

![organic2vs1](https://user-images.githubusercontent.com/35322900/170874973-bd6cf78b-1a28-4625-b25a-a772406f63e2.JPG)

slevisconsulting replied:
When I get a chance, I'm curious to understand this small difference. Meanwhile though I've confirmed what should go without saying, i.e. that ORGC, CFRAG, BULK, PHAQ (and all other fields) are identical between these two fsurdats.

slevisconsulting replied:
Oh and old fsurdat shows max organic = 130 as we discussed:

![organic_old](https://user-images.githubusercontent.com/35322900/170875584-747729f6-9a1d-4a73-8681-5be47df410ea.JPG)

dlawrenncar commented:
This looks encouraging.  In terms of trying to understand what the differences look like between the old fsurdat and the new data, perhaps you could plot with maximum value of 130, so we can see how the two datasets are similar/distinct.  There are big difference, for example in the Canadian Archipelago and around Hudson Bay that I can see.  No reason for me to think that the new data is wrong, though.  I'd also be interested to see how this looks going down into the soil.  That is, what do these datasets look like at 20cm, 50cm depth?

slevisconsulting commented:
> perhaps you could plot with maximum value of 130, so we can see how the two datasets are similar/distinct. There are big difference, for example in the Canadian Archipelago and around Hudson Bay that I can see. No reason for me to think that the new data is wrong, though.

New versus old layer 1, which goes to 1.75 cm:
![organic_new_vs_old](https://user-images.githubusercontent.com/35322900/171253698-9f871356-cf1e-48c1-8b45-1e7b99d75fa6.JPG)

> [...] what do these datasets look like at 20cm, 50cm depth?

Comparing at 16.6 cm
![organic_new_vs_old_16 6cm](https://user-images.githubusercontent.com/35322900/171258575-62561457-0478-421d-91ac-1e37e585e713.JPG)

Comparing at 28.9 cm
![organic_new_vs_old_28 9cm](https://user-images.githubusercontent.com/35322900/171258603-74906d07-5f50-4880-a99e-b46b788202c5.JPG)

Comparing at 49.3 cm
![organic_new_vs_old_49 3cm](https://user-images.githubusercontent.com/35322900/171258635-d5f7583a-9332-4471-ba0a-6bb835a8240d.JPG)

wwieder commented:
There are clear differences between datasets.  Maybe the best way to understand the impacts would be to run simulations that compare permafrost extent, soil temperatures, and hydrology / runoff.

slevisconsulting commented:
At our 2022/9/8 meeting we agreed on a year-2000 1-degree SP simulation with all the new raw datasets.
Question: How long of a simulation do we want?

To generate the fsurdat file for this simulation with the new mksurfdata_esmf:
- I'm updating the glacier data as discussed in #1406 
- I'm updating the lake data as discussed in #1106 

If I am not mistaken, two issues currently in progress will require pointing to new raw datasets:
- #1556 @lawrencepj1 
- #591 @olyson 

I may try the simulation before these issues complete and repeat it once they complete.

slevisconsulting commented:
As I have now run simulations with fsurdat files coming out of this branch, and we will likely run more,
@ekluzek recommends that I merge this branch to the ctsm5.2.mksurfdata branch and make a tag.

Once I do that, @ekluzek will merge the latest ctsm tag to the ctsm5.2.mksurfdata branch and make a tag.

slevisconsulting commented:
But first...
I was reviewing my code changes in this PR and found that we had been comparing two methods of obtaining ORGANIC. The preferred method was "option 1" rather than "option 2", which means that I must rerun the new_fsurdat simulation to use option 1. Once I'm satisfied with the results, I will push another commit to this PR and then merge to ctsm5.2.mksurfdata.

slevisconsulting commented on cime_config/testdefs/testmods_dirs/clm/smallville_dynlakes_monthly/user_nl_clm:
```diff
@@ -1,7 +1,7 @@
 
 ! This file was created with the following command:
-! ncap2 -s 'PCT_LAKE=array(0.0,0.0,PCT_CROP); PCT_LAKE={0.,50.,25.,25.,25.,25.}; HASLAKE=array(1.,1.,AREA); PCT_CROP=array(0.0,0.0,PCT_LAKE); PCT_CROP={0.,25.,12.,12.,12.,12.}' landuse.timeseries_1x1_smallvilleIA_hist_78pfts_simyr1850-1855_c160127.nc landuse.timeseries_1x1_smallvilleIA_hist_78pfts_simyr1850-1855_dynLakes_c200928.nc
+! ncap2 -s 'PCT_LAKE=array(0.0,0.0,PCT_CROP); PCT_LAKE={0.,50.,25.,25.,25.,25.}; PCT_LAKE_MAX=array(1.,1.,AREA); PCT_CROP=array(0.0,0.0,PCT_LAKE); PCT_CROP={0.,25.,12.,12.,12.,12.}' landuse.timeseries_1x1_smallvilleIA_hist_78pfts_simyr1850-1855_c160127.nc landuse.timeseries_1x1_smallvilleIA_hist_78pfts_simyr1850-1855_dynLakes_c200928.nc
```
@billsacks this is the only change in this commit where I didn't feel confident. Could you confirm? Other than the variable name, would you have also changed the array values for the purposes of this test?

slevisconsulting replied:
Thanks, @billsacks 
I will take care of this later since I merged this branch earlier today. 
Probably in a PR that will address #1703
